src/main/java/com/aws/greengrass/device/session/SessionManager.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.device.session;
+
+import com.aws.greengrass.device.exception.AuthenticationException;
+import com.aws.greengrass.device.iot.Certificate;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import lombok.AccessLevel;
+import lombok.Getter;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Singleton class for managing AuthN and AuthZ session.
+ */
+public class SessionManager {
+    private static final Logger logger = LogManager.getLogger(SessionManager.class);
+    private static final String SESSION_ID = "SessionId";
+
+    // TODO: Add mechanism to prevent session leaks
+    // This implementation currently relies on clients to properly close sessions. If they don't, then sessions
+    // will be leaked. We can add this once sessions are created with the entire set of device credentials. However,
+    // Moquette is currently creating sessions using just the certificate. Since multiple clients could use the same
+    // certificate, we can't de-dup based on this alone.
```

Comment:
what will we dedupe on in the future? Client id + cert?

If the cert is the identity, then is there actually a problem in using the same session since the authZ decisions would be the same?

Reply:
The cert won't be the identity. Moquette needs to be changed to use mqtt device credentials.

The current code "attaches" the thing after the fact for AuthZ decisions. You can just trust me on this one ;)